### DISCUSSION TITLE
Support product-specific content in editor field

### DISCUSSION THREAD

**Discussion by leewillis77** [woocommerce/woocommerce#48313] [WooCommerce New Product Editor]
Currently, the fields that are added to the product by an extension are specified without any context (unlike the current system where the currently edited product is available when added metaboxes etc). That means that the labels/description/helptext* are fixed at the point of definition and there isn't an easy way to provide product-specific content. 

Here's an example of one of our extension fields on two different products:

<img width="329" alt="Screenshot 2024-06-10 at 10 22 04" src="https://github.com/woocommerce/woocommerce/assets/1097338/48243c24-7184-4457-bf74-0f49451dae9d">
<img width="272" alt="Screenshot 2024-06-10 at 10 24 53" src="https://github.com/woocommerce/woocommerce/assets/1097338/549cbfd5-bf2d-4b33-9f4d-b00d8d11506d">

Notice how the "Default" text shows a different value. That's currently not possible with the new product editor wthout building all of the fields as custom components. It would be good if there was the ability to modify the label/description/help text on the standard components at runtime. 

* see also https://github.com/woocommerce/woocommerce/issues/43253 for a discussion about the inconsistency of what additional content is available on fields

**Comment by mattsherman** [comment]
@leewillis77 Thanks for starting this discussion.

In your example, how would you ideally like to specify the product-specific content in the labels, etc.? It seems like some sort of format string would could to be specified at field definition time, which could then be filled in at runtime, for simple cases. But, in your example, how is the different default being determined?

It seems that a simple format string wouldn't suffice, and some client-side code/expression would need to be executed at runtime. How would you want to specify that code? 

> if there was the ability to modify the label/description/help text on the standard components at runtime.

You can do this via a custom blocks, that uses the same underlying core WP components that the generic blocks do. The challenge will be to provide a mechanism to do this in a less complex manner than writing a custom block.

Maybe a client-side JS filter could be used to accomplish this. You'd have to write a little bit of JS, but it wouldn't be as much as writing an entire component.

**Comment by leewillis77** [comment]
> In your example, how would you ideally like to specify the product-specific content in the labels, etc.? It seems like some sort of format string would could to be specified at field definition time, which could then be filled in at runtime, for simple cases.

Unfortunately it's not quite that simple. Some fields will have no text, others will have a string in the format "Default: %s", others will have descriptions such as "Uses value from %s field if set.".

> But, in your example, how is the different default being determined?

The actual string shown depends on the core extension settings, and settings made against the categories that a product is in. I've already resigned myself that I'll need to provide a data store that provides the content for the fields for a given product. The issue is how then to take that content and have it displayed alongside the editor fields without having to create custom components for all of the fields.

I'm aware of several solutions which might work; slotFills, block hooks, JS filter for the label/help-text/description, but I don't know enough about the internals to say which is the "best", or "right" solution.
